### PR TITLE
[FIX] {sale,pos}_loyalty: compute points including discount lines

### PR DESCRIPTION
**Pull request by sbel-odoo**
Currently when applying a coupon for 10% on the order on the current order, the points awarded are 100% of the points instead of 90%. The loyalty program rewarding x points per $ spent.

The same behavior is also observed in the sale worflow.

Steps to reproduce:
-------------------
* Go to the **Point of Sale** App
* Open shop session
* Select a customer
* Add a product (ex. Office chair 70$)
* Select **Enter Code**
* Enter the code `10pc`
> Observation: The order will grant 700 points instead of 630. The amount paid is 63$ with the code for 10% on the order.

Settings of the **Discount & Loyalties**:
* **Code for 10% on orders**
  * Rule: 
    * `Among` -> no restriction set 
    * `Minimum quantity`: 1 
    * `Minimum Purchase`: 0
  * Reward:
    * `Discount`: 10% `on` order
* **Loyalty Program**
  * `Use points on`: Current & Future orders
  * Rule:
    * `Among` -> no restriction set
    * `Minimum quantity`: 1 
    * `Minimum Purchase`: 0 
    * `Grant` 10 loyalty poins per $ spent

Idea of the fix
---------------
As discussed with DALA, discount should count toward points awarded. Here is how it should apply:
* If program_rule has no domain, all lines should count (except ewallet & gift cards)
* If program_rule has a domain, we should use all the lines where the product matches the domain (except ewallet & gift cards)

This applies for the computation of points based on the money spent.

Why the fix:
------------
https://github.com/odoo/odoo/blob/2f6f8014f79e1db53d6c4c27e0cc60ccd48fd301/addons/sale_loyalty/models/sale_order.py#L917

Sale-wise, we replace `rule_products` with `so_products_per_rule.get(rule, [])`. We also replace `order_lines` by `lines_per_rule`.
* `rule_products` don't include discounts as they are computed on `order_lines` (all lines that are not reward lines). https://github.com/odoo/odoo/blob/2f6f8014f79e1db53d6c4c27e0cc60ccd48fd301/addons/sale_loyalty/models/sale_order.py#L834-L841 https://github.com/odoo/odoo/blob/2f6f8014f79e1db53d6c4c27e0cc60ccd48fd301/addons/sale_loyalty/models/sale_order.py#L889
* `so_products_per_rule` will include the discounts if they match with the program's domain.
https://github.com/odoo/odoo/blob/2f6f8014f79e1db53d6c4c27e0cc60ccd48fd301/addons/sale_loyalty/models/sale_order.py#L844
* We don't need to exclude the free product reward lines from `lines_for_points` as the price is 0 on the SO. Shipping lines and free shipping rewards are not counted.

In pos, the fix is a bit different.

https://github.com/odoo/odoo/blob/4b9dee8fbe9aefde798c8543fef89c661cc85c57/addons/pos_loyalty/static/src/js/Loyalty.js#L908-L914

We first notice that this piece of code is never triggered as the structure of `program.id` is a number and `reward.program_id` is the following thus the last if statement is never true. A `.id` is added.
```
Proxy(Object) {id: 1, name: 'Code for 10% on orders', trigger: 'with_code', applies_on: 'current', program_type: 'promo_code', …}
```

This now means that we only count reward lines if they are from the same program. 

With the following if statement, it means that only the lines that are not reward lines are counted (which excludes discounts).

https://github.com/odoo/odoo/blob/eadfa4ee290a8b62956097aa66989eacff230275/addons/pos_loyalty/static/src/js/Loyalty.js#L921-L924

We conclude that those two ifs must be changed as we want a similar behavior as the workflow in sales.

We remove `orderedProductPaid` out of the if statement since it can be impacted by reward lines. We change the first if to excludes the reward lines only from the current program or from ewallet and giftcards. With this, `orderedProductPaid` now counts the discount lines from other program. (Similar to sales, where `amount_paid` is computed with `lines_per_rule`)

`totalProductQty` still applies only on non reward lines. (Same as sales)

opw-3858286

### CODE REVIEW

**Comment by sbel-odoo** [comment]
Note for the reviewers: The test `test_points_awarded_specific_discount_code_specific_domain_program` is slightly different in POS than in sales as I also had to make the reward product available in the POS to have it in the `valid_product_ids`. It might not be the best to have to show discount products in the session to be able to configure the loyalty program. Do you think we could safely remove this condition when computing the `valid_product_ids`? https://github.com/odoo/odoo/blob/ce1c9d714b7760ce20e51df930da64cfe0b6efe2/addons/pos_loyalty/models/loyalty_rule.py#L29

File: addons/pos_loyalty/static/src/tours/PosLoyaltyTour.js
Comment (robinengels) [inline addons/pos_loyalty/static/src/tours/PosLoyaltyTour.js:373]: maybe add a step to check the point awarded in the PoS. Same for the other tours

**Review by robinengels — COMMENTED** [review]
LGTM for the point of sale part.

**Review by caburj — APPROVED** [review]
Impressive. This even took into account the past edge cases that was already fixed.

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @sbel-odoo :wave: 

I just tested your PR and I still witness the bug :/ Could you also check your test since it's green on runbot?

**Comment by sbel-odoo** [comment]
WIP. POS-wise nothing to say, Sale-wise coupons work but not promotions (global discounts) will investigate

File: addons/sale_loyalty/models/sale_order.py
Comment (chevalierv) [inline addons/sale_loyalty/models/sale_order.py:925]: Since we already loop over the lines, let's sum them directly :)
```suggestion
                        # NOTE: this accounts for discounts -> 1 point per $ * (100$ - 30%) will result in 70 points
                        lines_for_points = []
                        amount_paid = 0.0
                        rule = so_products_per_rule.get(rule, [])
                        for line in self.order_line - self._get_no_effect_on_threshold_lines():
                            is_discount = line.reward_id.reward_type == 'discount'
                            reward_program = line.reward_id.program_id
                            if line.reward_id and reward_program.program_type in ['ewallet', 'gift_card', program.program_type]:
                                continue
                            lines_for_points.append(line)
                            amount_paid += line.price_total if line.product_id in rule else 0.0
```

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @sbel-odoo 

Thanks for your investigation on this! Kudos for adding tests :tada:  

Have a nice day!

File: addons/sale_loyalty/models/sale_order.py
Comment (ltinel) [inline addons/sale_loyalty/models/sale_order.py:921]:
```suggestion
                            if line.reward_id.program_id.program_type in ['ewallet', 'gift_card', program.program_type]:
```

File: addons/sale_loyalty/tests/test_loyalty.py
Comment (ltinel) [inline addons/sale_loyalty/tests/test_loyalty.py:428]: Nit (same for the next test):
```suggestion
            'name': "Food",
```

**Review by ltinel — CHANGES_REQUESTED** [review]
Thanks for the fix, it looks good (and it wasn't all that easy to get right) ! :) Great test coverage BTW :+1:

File: addons/sale_loyalty/models/sale_order.py
Comment (sbel-odoo) [reply addons/sale_loyalty/models/sale_order.py:925]: Thank you ! I've changed the code locally to include this suggestion. I'll push it once I also include changes based on Itinel's review

File: addons/sale_loyalty/models/sale_order.py
Comment (sbel-odoo) [reply addons/sale_loyalty/models/sale_order.py:921]: Hope I can make it make sense :sweat_smile: 

Just to recall, at this point in the code we have the following:
- The rule of the program grants points https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L894-L895
- The program applies on either `current orders` or `current & future` and the points are not. We are in the else of this if condition. https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L897
- The program grans points per $ spent https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L914

Technically you can apply multiple coupons but not from the same program. I've tested the scenarios and I get a message like **"This promotion was already applied"**. Promotions and discount programs usually are set to grant 1 point per order if some conditions are met. So they do not really enter this part of the code.

This setting (what the program applies on + granting points per $ spent) is usually seen with the loyalty program type. This is also usually (functionally) the only program type that will have multiple different rewards.

> Would you mind explaining me why you're also skipping programs with the same type? 

Well this is mostly for the loyalty program. From the discussion I've had with DALA I understand that rewards that were obtained by spending loyalty points should not have the effect of diminishing the amount of points awarded. Here a more concrete example. Let's say you exchange 1000 loyalty points for a 10$ discount on your order. The order is simply 1 product of 50$. You should still get 50 * `nbr points per $`. 

I could write it like (with your last comment taken into account)
```python
if line.reward_id.program_id.program_type in ['ewallet', 'gift_card']:
    continue
if program.program_type == "loyalty" and line.reward_id.program_id.program_type == "loyalty"
```
This is more clearer to understand but maybe it should still include all programs. Let's say it's possible that through dev mode some configurations are so that there is a coupons program that would grant points per dollars spent if rule conditions are met. W just assume rule conditions are met. He gets 0.01 point per $ spent, he has an order of 100$, so he should get 1 point that he can use for the coupon (since coupons discount rewards are set in exchange of x points, usually 1). His order then is 90$ which means he should get 0.9 points and not 1 that he had spent.

This is a weird setting for discounts I can conceive but we've had customer that had loyalty programs that were like:
You get 1 point per $ spent and you can exchange 100 points for 1 $ on your orders. The program was set to applies on current and future orders. So for a 100$ order the clients should get 100 point that he can decide to use immediately but then his total is 99 and not 100 so he doesn't have the 100 point to spend immediately ( without the fix). This is the use case I discussed with DALA when he said that we should not deduct points twice (the use of points to get the reward and the deduction of points because of the reward).

Sorry if it's not clear. Feel free to message me for a call if you prefer :smile:

File: addons/sale_loyalty/models/sale_order.py
Comment (sbel-odoo) [reply addons/sale_loyalty/models/sale_order.py:918]: With my previous comment I explain that it can't be nested out of the program loop :smile: However it cannot be nested out of the rule loop if I apply Valentin's comment. So the question is do I remove the loop here 
```python
amount_paid = sum(line.price_total for line in lines_for_points if line.product_id in so_products_per_rule.get(rule, []))
``` 
This even remove the need to remember the `lines_for_points` or do I leave it and un-nest the original loop where I keep `lines_for_points` ?

File: addons/sale_loyalty/models/sale_order.py
Comment (sbel-odoo) [reply addons/sale_loyalty/models/sale_order.py:916]: I don't have acces to the article but no problem I can do it!

File: addons/sale_loyalty/tests/test_loyalty.py
Comment (sbel-odoo) [reply addons/sale_loyalty/tests/test_loyalty.py:54]: Sure I'll remember that from now :)

File: addons/sale_loyalty/tests/test_loyalty.py
Comment (sbel-odoo) [reply addons/sale_loyalty/tests/test_loyalty.py:83]: Sorry I've just re-used the method that exists in POS. I understand the first part about the single program but not so much that last part. I guess that you talk about using `cls.env['loyalty.program'].create()` but how will i be able to get the base setting configuration for the program type since it's not the same for all programs? 
If I simply do that: 
```python
`cls.env['loyalty.program'].create({
    'name': name;
    'program_type': program_type,
})`
```
it is enough?

**Comment by sbel-odoo** [comment]
@ltinel I'm pinging you to be sure you'll get a mail notification since I don't know if you'll get one with my comments. I answered some of your comments. I'll wait for you reply before pushing new code

File: addons/sale_loyalty/models/sale_order.py
Comment (ltinel) [reply addons/sale_loyalty/models/sale_order.py:921]: Got it, thanks for the detailed explanation ! :)

File: addons/sale_loyalty/models/sale_order.py
Comment (ltinel) [reply addons/sale_loyalty/models/sale_order.py:918]: It depends how many programs are filtered out by the condition in your loop. If there are not too many, then VCR's suggestion is a bit better IMO because it simplifies the code (and keeps the logic in one single place). And I suppose there are indeed not that many, as most people probably won't be using loads of Gift Cards and eWallets on a single order.

Would be nice to have some stats about this but it's not worth the effort IMO :sweat_smile:

@chevalierv WDYT?

File: addons/sale_loyalty/tests/test_loyalty.py
Comment (ltinel) [reply addons/sale_loyalty/tests/test_loyalty.py:83]: If the program's fields have no influence on the test, then I think you can indeed simply create a program with `name` and `program_type`. Alternatively, if you do need the base setup, you can do something like `LoyaltyProgram.create(LoyaltyProgram._get_template_values()[program_type])` (which does the same as what you have, but without creating an action).

If you do rely on some of the program's fields, then I would set them explicitly (so that it's clear where they come from, and so that your test doesn't break if someone changes them).

**Comment by sbel-odoo** [comment]
Since PR is getting long I pushed a fixup to make sure everything is ok. Will squash once approved

File: addons/sale_loyalty/tests/test_loyalty.py
Comment (ltinel) [inline addons/sale_loyalty/tests/test_loyalty.py:338]: Nit: is you use an additional variable, you could fit this on a single line :)
```suggestion
        LoyaltyProgram = self.env['loyalty.program']
        loyalty_program = LoyaltyProgram.create(LoyaltyProgram._get_template_values()['loyalty'])
```

**Review by ltinel — APPROVED** [review]

File: addons/sale_loyalty/models/sale_order.py
Comment (chevalierv) [reply addons/sale_loyalty/models/sale_order.py:949]: Indeed, I guess that it's unlikely to have many programs here

**Review by chevalierv — APPROVED** [review]
@robodoo delegate+

**Comment by sbel-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
